projects/RabbitMQ.Client/client/impl/Frame.cs

```diff
@@ -169,99 +152,88 @@ private InboundFrame(FrameType type, int channel, ReadOnlyMemory<byte> payload,
-                // ProtocolViolationException, doesn't permit
-                // this. Fortunately, the call stack in the
-                // EndOfStreamException is largely irrelevant at this
-                // point, so can safely be ignored.
                 throw new MalformedFrameException("Invalid AMQP protocol header from server");
             }
+
+            int transportHigh = protocolError[3];
+            int transportLow = protocolError[4];
+            int serverMajor = protocolError[5];
+            int serverMinor = protocolError[6];
+            throw new PacketNotRecognizedException(transportHigh, transportLow, serverMajor, serverMinor);
         }
 
-        internal static InboundFrame ReadFrom(Stream reader, byte[] frameHeaderBuffer)
+        internal static bool TryReadFrame(ref ReadOnlySequence<byte> buffer, out InboundFrame frame)
         {
-            int type = default;
-            try
+            // We'll always need to read at least 8 bytes (type (1) + channel (2) + payloadSize (4) + end marker (1)) or (8 bytes of protocol error, see ProcessProtocolHeader).
+            if (buffer.Length < 8)
             {
-                type = reader.ReadByte();
+                frame = default;
+                return false;
             }
-            catch (IOException ioe)
+
+            if (buffer.First.Span[0] == 'A')
             {
-                // If it's a WSAETIMEDOUT SocketException, unwrap it.
-                // This might happen when the limit of half-open connections is
-                // reached.
-                if (ioe.InnerException is null ||
-                    !(ioe.InnerException is SocketException exception) ||
-                    exception.SocketErrorCode != SocketError.TimedOut)
+                // Probably an AMQP protocol header, otherwise meaningless
+                if (buffer.First.Length >= 8)
                 {
-                    throw;
+                    ProcessProtocolHeader(buffer.First.Span.Slice(1, 7));
+                }
+                else
+                {
+                    Span<byte> protocolError = stackalloc byte[7];
+                    buffer.Slice(1, 7).CopyTo(protocolError);
+                    ProcessProtocolHeader(protocolError);
                 }
-
-                ExceptionDispatchInfo.Capture(ioe.InnerException).Throw();
             }
 
-            switch (type)
+            int type = buffer.First.Span[0];
+            int channel;
+            int payloadSize;
+
+            if (buffer.First.Length >= 7)
             {
-                case -1:
-                    throw new EndOfStreamException("Reached the end of the stream. Possible authentication failure.");
-                case 'A':
-                    // Probably an AMQP protocol header, otherwise meaningless
-                    ProcessProtocolHeader(reader);
-                    break;
+                channel = NetworkOrderDeserializer.ReadUInt16(buffer.First.Span.Slice(1, 2));
```

Comment:
there are alos extension methods on the sequence reader like 
https://docs.microsoft.com/en-us/dotnet/api/system.buffers.sequencereaderextensions.tryreadbigendian?view=netcore-3.1

so